Added DebugInput Singleton

savage13 opened the pull request:
Added DebugInput Singleton

I was unsure of where to place it, so it is in `src/Game`

Thanks all for the help on this :D

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/zeldaret/botw/72)
<!-- Reviewable:end -->

leoetlino commented on src/Game/CMakeLists.txt:
```diff
@@ -9,4 +9,6 @@ target_sources(uking PRIVATE
   gameScene.h
   gameStageInfo.cpp
   gameStageInfo.h
+  debugInput.cpp
```
Could you move this to KingSystem (ksys)? Looks like DebugInput is initialised from ksys::InitializeApp, which means it is part of the framework part of BotW (`src/KingSystem`), not the game part (`src/Game`). Not sure where it should go -- maybe just put it in `KingSystem/ksysDebugInput.cpp` (unless you see a better subfolder in KingSystem?)

leoetlino commented on src/Game/debugInput.cpp:
```diff
@@ -0,0 +1,4 @@
+
+#include "debugInput.h"
```
Include paths should be "project-absolute", i.e. relative to `src/`. If the .h is in `src/KingSystem/ksysDebugInput.h` then the include should be `#include "KingSystem/ksysDebugInput.h"`.

leoetlino commented on src/Game/debugInput.cpp:
```diff
@@ -0,0 +1,4 @@
+
```
There's an extra new line that should be removed.

leoetlino commented on src/Game/debugInput.h:
```diff
@@ -0,0 +1,16 @@
+#include <heap/seadDisposer.h>
+
+class DebugInput {
+    SEAD_SINGLETON_DISPOSER(DebugInput);
```
```suggestion
    SEAD_SINGLETON_DISPOSER(DebugInput)
```

leoetlino commented on src/Game/debugInput.h:
```diff
@@ -0,0 +1,16 @@
+    virtual ~DebugInput() = default;
+
+private:
+    bool t1 = false;
```
To keep things consistent and obvious, the convention we usually use for placeholder names is `_offset_in_lowercase_hex`, so this would be `bool _8;`

(You may be wondering why the bool is at offset 8 and not offset 0: that's because the virtual function table pointer ("vtable pointer") is at offset 0.)

leoetlino commented on src/Game/debugInput.h:
```diff
@@ -0,0 +1,16 @@
+
+private:
+    bool t1 = false;
+    u32 t2 = 0;
```
```suggestion
    u32 _c = 0;
```
(Note that this is not at offset 8+1 = 9 because of alignment reasons: uint32s are usually 4-byte aligned. The compiler automatically inserts invisible padding bytes to make fields correctly aligned.)

leoetlino commented on src/Game/debugInput.h:
```diff
@@ -0,0 +1,16 @@
+
+#include <heap/seadDisposer.h>
+
+class DebugInput {
```
Missed this earlier, but this should be in the `ksys` namespace

savage13 commented:
- Moved `src/Game/DebugInput.[h.cpp]` into `src/KingSystem/ksysDebugInput.[h,cpp]`
- Added `ksys` namespace
- Renamed private variables
- Not sure if I got the extra newline :|

savage13 commented on data/uking_functions.csv:
```diff
@@ -94369,7 +94369,7 @@ Address,Quality,Size,Name
 0x0000007101266f6c,U,000108,
-0x0000007101266fd8,U,000152,DebugInput::createInstance
+0x0000007101266fd8,O,000152,_ZN4ksys10DebugInput14createInstanceEPN4sead4HeapE
 0x0000007101267070,U,000160,
```
If you have time, I would be willing to try

leoetlino commented on data/uking_functions.csv:
```diff
@@ -94369,8 +94369,8 @@ Address,Quality,Size,Name
 0x0000007101266ecc,U,000060,
 0x0000007101266f08,U,000100,
```
These two functions might be the DebugInput::SingletonDisposer destructor functions (which you get for free when you use the SINGLETON macro). 

You can use `tools/listsym DebugInput` to figure out their names and then list them

savage13 replied:
```
% ./tools/check -mw _ZN4ksys10DebugInput18SingletonDisposer_D0Ev
ksys::DebugInput::SingletonDisposer_::~SingletonDisposer_() (_ZN4ksys10DebugInput18SingletonDisposer_D0Ev)
OK
note: changing status from NotDecompiled to Matching

% ./tools/check -mw _ZN4ksys10DebugInput18SingletonDisposer_D1Ev
ksys::DebugInput::SingletonDisposer_::~SingletonDisposer_() (_ZN4ksys10DebugInput18SingletonDisposer_D1Ev)
OK
```

```
0x0000007101266ecc,U,000060,
0x0000007101266f08,O,000100,_ZN4ksys10DebugInput18SingletonDisposer_D1Ev
0x0000007101266f6c,O,000108,_ZN4ksys10DebugInput18SingletonDisposer_D0Ev
0x0000007101266fd8,O,000152,_ZN4ksys10DebugInput14createInstanceEPN4sead4HeapE
0x0000007101267070,O,000160,_ZN4ksys10DebugInput6updateEv
```

leoetlino commented on src/KingSystem/ksysDebugInput.h:
```diff
@@ -0,0 +1,21 @@
+    void update();
+
+private:
+    sead::BitFlag8 mFlags = false;
```
don't forget the include for BitFlag8 (also you can remove `= false`, BitFlags are automatically initialised to 0)

leoetlino reviewed (approved):
lgtm, the only (extremely minor) "issues" left are the two new extra blank lines at the top that should be removed

leoetlino commented:
One last thing: could you squash/"fixup" your commits please? I usually don't ask it for objmap PRs, but in this repo, PRs tend to get larger so a clean, unmangled commit history is especially important!

```
git remote add upstream https://github.com/zeldaret/botw
git fetch upstream
git rebase -i upstream/master
```

then change "pick" into "f" or "fixup" for all except the commit whose message you want to keep

savage13 commented:
I think I did the rebase/squash incorrectly

savage13 commented:
I am closing this and will push a clean pull request.
